### PR TITLE
refactoring code, changing halls actions to services and adding tests

### PR DESCRIPTION
**Pull request by danyflorczak**
- small refactors in movie controllers
- display changes for movies index
- adding create, update, delete services for halls 
- adding tests to those services

### CODE REVIEW

**Comment by danyflorczak** [comment]
Now that I saw how big this pr I know that it is definitely too big and I'll try to remember about it while refactoring and split it into much smaller pr's

File: spec/services/halls/update_spec.rb [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  let(:params) do
+    {
+      name:,
+      capacity: 41
+    }
+  end
+  let(:name) { 'name updated' }
+  let(:instance) { described_class.new(hall.id, params) }
+
+  subject { instance.call }
```
Comment (wijanko) [inline spec/services/halls/update_spec.rb]:
```suggestion
  subject(:hall_update) { described_class.new(hall.id, params).call }
```

`instance` it not used in any other place, you can remove it.
Also you can name you subject, then you can read test scenarios like in my other comment.

File: spec/services/halls/update_spec.rb [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+RSpec.describe Halls::Update do
+  let!(:hall) { create :hall, name: 'name', capacity: 40 }
+  let(:params) do
+    {
+      name:,
+      capacity: 41
+    }
+  end
+  let(:name) { 'name updated' }
```
Comment (wijanko) [inline spec/services/halls/update_spec.rb]:
```suggestion
  let(:name) {  Faker::Name.name }
```
I recommend using faker, here you have some available [methods](https://github.com/faker-ruby/faker#default)

File: spec/services/halls/delete_spec.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe Halls::Delete do
+  let!(:hall) { create :hall, name: 'name', capacity: 30 }
```
Comment (wijanko) [inline spec/services/halls/delete_spec.rb]:
```suggestion
  let!(:hall) { create :hall }
```

name and capacity has no special role in this test, you can use [defaults](https://github.com/danyflorczak/Monte_Cinema/pull/16/files#diff-e1e712fa419ca15713587465c1ad6a27f6235af650bdfe84b049f28574de068cR5-R6).

File: spec/services/halls/create_spec.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+require 'rails_helper'
+
+RSpec.describe Halls::Create do
+  let(:params) do
+    {
+      name: 'name',
+      capacity: 100
+    }
+  end
+  let(:instance) { described_class.new(params) }
```
Comment (wijanko) [inline spec/services/halls/create_spec.rb]:
```suggestion
  subject { described_class.new(params).call }
```

In the other test cases you use subject, please stick to one convention.

File: spec/services/halls/update_spec.rb [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  describe '.call' do
+    it 'updates hall' do
+      expect { subject }
+        .to change { hall.reload.name }.from('name').to('name updated')
+        .and change { hall.reload.capacity }.from(40).to(41)
+    end
+
+    it 'retuns hall object' do
+      expect(subject).to eq(hall)
+    end
```
Comment (wijanko) [inline spec/services/halls/update_spec.rb]:
```suggestion
    it 'retuns hall object' do
      expect(hall_update).to eq(hall)
    end
```

also you have some other shorter syntax if you declared subject

```suggestion
    it { is_expected.to eq(hall) }
```

File: spec/factories/hall.rb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :hall do
+    name { 'My name' }
+    capacity { 20 }
```
Comment (wijanko) [inline spec/factories/hall.rb]:
```suggestion
    capacity { Faker::Number.number(digits: 2).to_i }
```

You can use Faker for all the factiories unless you need to have specific value.
I gave you example with casting numbers to integer.

File: app/services/reservations/confirm_reservation.rb [hunk 7]
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+module Reservations
+  class ConfirmReservation
+    def initialize(reservation)
+      @reservation = reservation
+    end
+
+    def call
+      return false unless booked?
```
Comment (wijanko) [inline app/services/reservations/confirm_reservation.rb]:
```suggestion
      return false unless reservation.booked?
```

You have reservation status declared as an enum so you can use an advantages of this approach and remove your helper [methods](https://api.rubyonrails.org/v7.0.3.1/classes/ActiveRecord/Enum.html).

File: app/services/halls/update.rb [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+module Halls
+  class Update
+    def initialize(hall_id, params)
+      @hall_id = hall_id
+      @params = params
+    end
+
+    def call
+      hall = Hall.find(hall_id)
```
Comment (wijanko) [inline app/services/halls/update.rb:11]: What if there will be no hall with certain id?

File: app/controllers/halls_controller.rb [hunk 9]
```diff
@@ -24,8 +24,9 @@ def edit
...
   def create
     authorize Hall
-    @hall = Hall.new(hall_params)
-    if @hall.save
+    @hall = ::Halls::Create.new(hall_params).call
+
+    if @hall.valid?
       redirect_to hall_url(@hall), notice: 'Hall was successfully created.'
     else
       render :new, status: :unprocessable_entity
```
Comment (wijanko) [inline app/controllers/halls_controller.rb:32]: Do you plan to pass error messages and inform users about the reasons of failures?

File: app/services/halls/delete.rb [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+module Halls
+  class Delete
+    def initialize(hall_id)
+      @hall_id = hall_id
+    end
+
+    def call
+      hall = Hall.find(hall_id)
+      hall.destroy
+    end
```
Comment (wijanko) [inline app/services/halls/delete.rb:12]:
```suggestion
    def call(hall_id)
      hall = Hall.find(hall_id)
      hall.destroy
    end
```

In such a short classes you can pass a `hall_id` to call method and remove all other things in this class.
You can think it over and make your decision.

**Review by wijanko — APPROVED** [review]
I added some comments in specific lines, but you can apply advices also in other similar places in this review.
I would summarize it as:
- use Faker when you can,
- use enum features if you have possibility,
- try to be consistent in tests, pay attention to things which are important for certain test and make them visible, also do not show data which wont affect current test scenarios

Overall I found it correct, good job 😉 
Approved 🎆

**Pull request merged**